File: HanJaehee/programmers/43164/programmers43164.java
import java.util.*;

class Solution {
    List<String> answer = new ArrayList<>();
    public String[] solution(String[][] tickets) {

        dfs(tickets, 0, "ICN", "ICN", new boolean[tickets.length]);
        Collections.sort(answer);
        return answer.get(0).split(" ");
    }
    public void dfs(String[][] tickets, int idx, String now, String total, boolean[] visit){
        if(idx == tickets.length){
            answer.add(total);
            return;
        }
        
        for(int i=0; i< tickets.length; ++i){
            if(!visit[i] && tickets[i][0].equals(now)){
                visit[i] = true;
                dfs(tickets, idx+1, tickets[i][1], total + " " + tickets[i][1], visit);
                visit[i] = false;
            }
        }
    }
    
}

File: swanious/programmers/43164/programmers_43164.py
def solution(tickets):
    l = len(tickets)
    visit = [False] * l
    ans = []
    
    tickets.sort()
    
    def dfs(start, cnt):
        global answer
        ans.append(start)
        
        if cnt == l: 
            return True
        
        for i in range(l):
            if not visit[i] and tickets[i][0] == start:
                visit[i] = True
                if dfs(tickets[i][1], cnt + 1): return True
                visit[i] = False
                
        ans.pop()
        return False
    
    dfs('ICN', 0)
    return ans